### ISSUE TITLE
[Meta][Trust Model] Case Study: Manifest-First Bypass via Interface Widening Without Spec Synchronization

### ISSUE THREAD

**Issue by Ibuki-wind** [tile-ai/TileOPs#839]
## Description

This issue records PR #834 as a concrete trust-model breach case study for spec-governed interface evolution under the TileOPs manifest workflow.

The goal is not to relitigate the PR. The goal is to preserve the exact failure mode in a reusable, self-contained form for future trust-model design, validator design, CI design, issue writing, and code review.

## Context

TileOPs defines a manifest-first trust model:
- [`docs/trust-model.md`](docs/trust-model.md) states that `ops_manifest.yaml` is the source of truth for op interfaces and that manifest changes are human-reviewed in a separate PR.
- [`docs/manifest.md`](docs/manifest.md) states that the manifest is the sole source of truth for op interfaces.
- [`docs/manifest.md`](docs/manifest.md) rule R15 requires manifest signatures to match PyTorch public API semantics.
- [`manifest-spec.md`](.claude/domain-rules/manifest-spec.md) says params include all PyTorch-supported parameters even if the current kernel only supports the default.

Under that model, the intended path for interface widening is:

1. Determine the PyTorch/public API contract.
2. Update `tileops/ops_manifest.yaml` in a separate manifest PR.
3. Have implementation and tests conform to the manifest.
4. Let CI/validator verify that code still matches the manifest.

PR #834 reversed that order.

## Incident Summary

PR #834 introduced `dim=None` support across reduction ops in implementation and tests before the manifest was updated.

Primary implementation commit:
- `993ec54` — `[Feat][Reduction] Support dim=None in all reduction ops for full reduction`

The commit message explicitly says:
- "Manifest changes deferred to a separate PR per trust model."

That sentence is the key breach marker: the implementation knowingly widened the public interface while the declared source of truth remained unchanged.

This produced a spec/implementation split:
- code accepted `dim=None`
- tests validated `dim=None`
- manifest still declared the old interface
- CI and validator did not block the drift

### 1. Manifest remained on the old interface contract

At PR #834 head (`c6fc350`), the manifest still declared reduction params without `None`:

- `logsumexp_fwd.params.dim`: `"int | list[int]"`
- `sum_fwd.params.dim`: `"int | list[int]"`
- `mean_fwd.params.dim`: `"int | list[int]"`
- `amax_fwd.params.dim`: `"int | list[int]"`
- `amin_fwd.params.dim`: `"int | list[int]"`
- `var_fwd.params.dim`: `"int | list[int]"`
- `std_fwd.params.dim`: `"int | list[int]"`
- `var_mean_fwd.params.dim`: `"int | list[int]"`
- `all_fwd.params.dim`: `"int | list[int]"`
- `any_fwd.params.dim`: `"int | list[int]"`
- `count_nonzero_fwd.params.dim`: `"int | list[int]"`
- `l1_norm_fwd.params.dim`: `"int | list[int]"`
- `l2_norm_fwd.params.dim`: `"int | list[int]"`
- `inf_norm_fwd.params.dim`: `"int | list[int]"`
- `prod_fwd.params.dim`: `int`

Representative references in the current tree:
- [`tileops/ops_manifest.yaml` lines 1409-1410](tileops/ops_manifest.yaml)
- [`tileops/ops_manifest.yaml` lines 1451-1452](tileops/ops_manifest.yaml)
- [`tileops/ops_manifest.yaml` lines 1599-1600](tileops/ops_manifest.yaml)
- [`tileops/ops_manifest.yaml` lines 1839-1840](tileops/ops_manifest.yaml)
- [`tileops/ops_manifest.yaml` lines 1911-1911](tileops/ops_manifest.yaml)
- [`tileops/ops_manifest.yaml` lines 2025-2026](tileops/ops_manifest.yaml)

### 2. Implementation widened the interface anyway

At the same PR head, implementation accepts `None` directly in public constructors and routes it through the reduction path:

- [`tileops/ops/reduction/reduce.py`](tileops/ops/reduction/reduce.py)
  - `_ReduceOpBase.__init__`: `dim: Union[int, List[int], None] = -1`
- [`tileops/ops/reduction/logsumexp.py`](tileops/ops/reduction/logsumexp.py)
  - `LogSumExpOp.__init__`: `dim: Union[int, List[int], None] = -1`
- [`tileops/ops/reduction/all_op.py`](tileops/ops/reduction/all_op.py)
  - `AllOp.__init__`: `dim: Union[int, List[int], None] = -1`
- [`tileops/ops/reduction/any_op.py`](tileops/ops/reduction/any_op.py)
  - `AnyOp.__init__`: `dim: Union[int, List[int], None] = -1`
- [`tileops/ops/reduction/count_nonzero.py`](tileops/ops/reduction/count_nonzero.py)
  - `CountNonzeroOp.__init__`: `dim: Union[int, List[int], None] = -1`
- [`tileops/ops/reduction/l1_norm.py`](tileops/ops/reduction/l1_norm.py)
  - `L1NormOp.__init__`: `dim: Union[int, List[int], None] = -1`
- [`tileops/ops/reduction/l2_norm.py`](tileops/ops/reduction/l2_norm.py)
  - `L2NormOp.__init__`: `dim: Union[int, List[int], None] = -1`
- [`tileops/ops/reduction/inf_norm.py`](tileops/ops/reduction/inf_norm.py)
  - `InfNormOp.__init__`: `dim: Union[int, List[int], None] = -1`

The core helper also treats `None` as a valid reduction specification:
- [`tileops/ops/reduction/_multidim.py`](tileops/ops/reduction/_multidim.py)
  - `normalize_dim(None, ndim) -> list(range(ndim))`

### 3. Tests validated the widened implementation, not the manifest contract

PR #834 added a dedicated file:
- [`tests/ops/test_reduce_dim_none.py`](tests/ops/test_reduce_dim_none.py)

This suite proves:
- `SumOp(..., dim=None)` matches `torch.sum`
- `MeanOp(..., dim=None)` matches `torch.mean`
- `VarOp(..., dim=None)` matches `torch.var`
- `LogSumExpOp(..., dim=None)` matches `torch.logsumexp`
- `AllOp/AnyOp/CountNonzeroOp/L1/L2/InfNorm(..., dim=None)` match the corresponding PyTorch semantics

That is valuable correctness evidence, but it validates:
- PyTorch semantics -> implementation

It does not validate:
- manifest -> implementation

So the test layer reinforced the widened runtime contract even while the declared interface source of truth still disagreed.

### 4. CI did not run the validator level that would even attempt interface checks

PR #834 head workflow configuration:
- [` .github/workflows/preflight.yml` lines 166-167](.github/workflows/preflight.yml)

Current command:
```yaml
python scripts/validate_manifest.py --levels schema,shape,dtype,bench
```

The `signature` level is absent.

Therefore, the preflight job never even attempted to compare manifest signatures against code for this PR.

### 5. Even if `signature` had run, the current validator still would not catch this class of drift

Current validator logic:
- [`scripts/validate_manifest.py` `check_l1_signature()`](scripts/validate_manifest.py)

The function checks only:
1. whether manifest-declared params exist by name in `__init__()` or `forward()`
2. whether `forward()` parameter order matches manifest order

It does **not** check:
- whether a Python parameter type is wider than the manifest type
- whether `None` was newly admitted
- whether defaults changed semantically
- whether manifest `int` is being implemented as `int | list[int] | None`

Concrete proof:

Running the validator on PR #834 head with `signature` enabled still passes:

```bash
python scripts/validate_manifest.py --levels schema,signature,shape,dtype,bench --check-op sum_fwd
python scripts/validate_manifest.py --levels schema,signature,shape,dtype,bench --check-op prod_fwd
python scripts/validate_manifest.py --levels schema,signature,shape,dtype,bench --check-op logsumexp_fwd
```

Observed result on the PR head worktree:
- all three commands passed
- `sum_fwd` passed despite manifest `int | list[int]` vs code `int | list[int] | None`
- `prod_fwd` passed despite manifest `int` vs code `int | list[int] | None`
- `logsumexp_fwd` passed despite manifest `int | list[int]` vs code `int | list[int] | None`

This shows the validator currently enforces only name/order conformance, not type/default/semantic conformance.

## Core Failure Mode

The trust model says:
- manifest is the source of truth
- implementation must not modify manifest in the same PR
- manifest changes require separate human review

But the actual executable system allowed the following path:

1. Read PyTorch API semantics directly.
2. Widen implementation to match PyTorch.
3. Add tests that validate the widened implementation against PyTorch.
4. Leave manifest unchanged because trust model says manifest must be separate.
5. Pass CI because CI does not run `signature` checks.
6. Still pass validator locally even with `signature` enabled because the validator only checks parameter names/order, not type widening.

That path breaks manifest-first governance while remaining green on the visible acceptance surface.

## Why This Is A Trust-Model Break, Not Just “Manifest Debt”

This was not a harmless delay in spec synchronization.

The system explicitly allowed a new public interface contract to become true in code and tests before it became true in the declared source of truth.

The breach is therefore:
- the manifest ceased to be the operative source of truth
- PyTorch semantics became the operative source of truth for the implementation path
- the validator/CI stack failed to detect the substitution

That is a direct violation of the intended trust chain.

## Exploit Path / Break Path

This case exposes a reusable break path for future agents or contributors:

1. Find a PyTorch-facing interface gap where the manifest lags the public API.
2. Widen the implementation directly to match PyTorch.
3. Add correctness tests against PyTorch.
4. Leave manifest untouched, citing separate-manifest-PR policy.
5. Rely on current CI configuration omitting `signature`.
6. Even if questioned, rely on current validator semantics, which only check parameter names/order and not type widening.

This path is dangerous because it is locally rational and looks compliant:
- it respects PyTorch semantics
- it keeps manifest edits separate
- it improves correctness relative to PyTorch
- tests pass
- validator passes
- CI passes

But it silently inverts the trust model.

## Commit Timeline

From PR #834:

- `993ec54` (2026-04-07 23:26:42 +08:00)
  - implementation + tests for `dim=None`
  - commit body explicitly says manifest changes are deferred to a separate PR
- `1aa3662` (2026-04-07 23:35:59 +08:00)
  - unrelated benchmark fix added on top
- `1d2613e` (2026-04-08 11:29:07 +08:00)
  - benchmark fix reverted after review feedback
- `c6fc350` (2026-04-08 11:32:19 +08:00)
  - test coverage expanded after review feedback

The benchmark scope leak was corrected. The manifest/implementation drift remained.

### 1. “Separate manifest PR” is not sufficient by itself

If the implementation is allowed to ship interface widening before the manifest PR lands, the manifest is no longer the operative source of truth.

The process needs an explicit rule such as:
- implementation PRs may not widen a public interface unless the corresponding manifest PR is already merged or explicitly referenced as the approved upstream contract

### 2. CI must run signature validation

The current preflight job omits `signature`. That creates a direct bypass channel.

### 3. Signature validation must check type/default/semantic conformance, not only names

A manifest validator that only checks parameter names/order cannot defend a manifest-first trust model.

It must also detect at least:
- added `None`
- widened unions
- default value drift
- parameter presence mismatches between manifest and public constructor contract

### 4. Tests against PyTorch are not a substitute for manifest conformance

PyTorch-reference tests verify behavioral correctness.
They do not verify spec governance.

Both checks are needed.

### 5. Public-API widening is a high-risk trust-boundary event

Any change that expands accepted user inputs should be reviewed as a spec event first, not merely as an implementation convenience.

## Proposed Follow-up

Use this case study as a standing reference for:
- manifest-first issue writing
- validator design for interface conformance
- CI design for manifest enforcement
- review guidance on public-API widening
- agent instructions when PyTorch semantics and manifest state diverge

## References

- PR #834
- Issue #814
- Issue #789
- [`docs/trust-model.md`](docs/trust-model.md)
- [`docs/manifest.md`](docs/manifest.md)
- [`scripts/validate_manifest.py`](scripts/validate_manifest.py)
